### PR TITLE
feat: state access/update gas

### PR DESCRIPTION
**Pull request by Stebalien**
Implement the state/address load/store gas charging logic described in FIP-0057.

fixes https://github.com/filecoin-project/ref-fvm/issues/1346

### CODE REVIEW

**Comment by Stebalien** [comment]
Unfortunately, this is definitely a consensus-critical change. My plan is to pull in two versions of the FVM from the hyperspace FFI branch instead of trying to add any checks here.

File: fvm/src/history_map.rs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+use std::borrow::Borrow;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::hash::Hash;
+
+/// A map with an "undo" history. All changes to this map are recorded in the history and can be "reverted" by calling `rollback`. Specifically:
+///
+/// 1. The user can call `history_len` to record the current history length.
+/// 2. The user can _later_ call `rollback(previous_length)` to rollback to the state in step 1.
+pub struct HistoryMap<K, V> {
```
Comment (Stebalien) [inline fvm/src/history_map.rs:12]: Moved to a shared module.

File: fvm/src/history_map.rs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    }
+}
+
+impl<K, V> HistoryMap<K, V>
+where
+    K: Hash + Eq + Clone,
+    V: Eq,
+{
+    /// Insert a k/v pair into the map, recording the previous value in the history if it differs.
+    pub fn insert(&mut self, k: K, v: V) {
```
Comment (Stebalien) [inline fvm/src/history_map.rs:44]: Updated to avoid inserting duplicate items.

**Comment by Stebalien** [comment]
Reviews:

- @arajasek for the full review.
- @Kubuxu in case he wants to take a look to make sure we're not doing anything too horrible.
- @raulk FYI.

File: fvm/src/call_manager/state_access_tracker.rs [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+use crate::kernel::{ClassifyResult, Result};
+
+struct StateAccessLayer {
+    addresses_height: usize,
+    actors_height: usize,
+}
+
+pub struct StateAccessTracker {
+    price_list: &'static PriceList,
+    actors: RefCell<HistoryMap<ActorID, bool>>,
```
Comment (arajasek) [inline fvm/src/call_manager/state_access_tracker.rs:21]: Comment explaining this `bool` value please (false => read-but-not-written, true => written)

File: fvm/src/call_manager/default.rs [hunk 4]
```diff
@@ -388,12 +507,23 @@ where
         s.exec_trace.push(trace);
     }
 
-    fn create_account_actor<K>(&mut self, addr: &Address) -> Result<ActorID>
+    /// Helper method to create an uninitialized actor due to a send.
+    fn create_actor_from_send(&mut self, addr: &Address, act: ActorState) -> Result<ActorID> {
+        let _ = self.charge_gas(self.price_list().on_create_actor(true))?;
+        let addr_id = self.state_tree_mut().register_new_address(addr)?;
+        self.state_access_tracker.record_lookup_address(addr);
```
Comment (arajasek) [inline fvm/src/call_manager/default.rs:514]: Why do we not record this as an Actor Update?

**Review by arajasek — COMMENTED** [review]
First pass review: This broadly looks good to me. I need to comb over more of it (eg. making sure all the new numbers actually match the FIP), but some notes for now.

I have 2 high-level questions:

- Regarding _where_ some of the logic lives:
  - Do we think the state access tracker would be cleaner if it only indicated _whether_ the relevant operation was cached (and if so at what "level" -- read vs. write), leaving it to the call manager impl itself to charge as appropriate? I don't think this is necessarily better, but it would make the logic of the `StateAccessTracker` much more scoped.
  - I'm not the _biggest_ fan of the "blessed" actors (singletons, burn, etc.) being in the price list. But I do think on balance it belongs there (and not in the CallManager), so not recommending a change here, just wanted to bring it up.
- This is a question for the FIP itself, but I wanted to make sure we're okay with the zero-gas charges for cached entries. There is still _some_ work happening, and I wanted to be sure that we're confident that some other cost is sufficient to protect (eg. the base syscall cost). What happens, for instance, if a contract just repeatedly resolves the same address?

**Comment by Stebalien** [comment]
> This is a question for the FIP itself, but I wanted to make sure we're okay with the zero-gas charges for cached entries. There is still some work happening, and I wanted to be sure that we're confident that some other cost is sufficient to protect (eg. the base syscall cost). What happens, for instance, if a contract just repeatedly resolves the same address?

My assumption is that the base syscall gas should be "enough", but we should add more benchmarks.

I quickly checked this against transfer. Overall, transfer costs ~6000 to update two states. That's 3k gas per update in the state-tree cache if I ignore the math involved. The syscall cost is 14000 gas, so this shouldn't be bad enough to actually be a security concern no matter what happens.

But yeah, I'd like to spend some time benchmarking later.

File: fvm/src/call_manager/default.rs (hunk 4 above)
Comment (Stebalien) [reply fvm/src/call_manager/default.rs:514]: 1. `on_create_actor` charge only charges for storage and address assignment.
2. `set_actor` below will charge to actually _create_ the actor (possibly charging to read it as well).

File: fvm/src/call_manager/state_access_tracker.rs (hunk 3 above)
Comment (Stebalien) [reply fvm/src/call_manager/state_access_tracker.rs:21]: I'm just going to use an enum.

File: fvm/src/call_manager/default.rs [hunk 5]
```diff
@@ -347,19 +382,20 @@ where
...
+                act
             }
             // Don't replace anything else.
             Some(_) => {
                 return Err(syscall_error!(Forbidden; "Actor address already exists").into());
             }
             // Create a new actor.
-            None => (ActorState::new_empty(code_id, delegated_address), true),
+            None => {
+                let _ = self.charge_gas(self.price_list().on_create_actor(false))?;
```
[the remark this answers is not in the document]
Comment (Stebalien) [reply fvm/src/call_manager/default.rs:393]: The address has already been assigned by the init actor (charging gas the _normal_ way). I'll add a comment.

File: fvm/src/call_manager/default.rs [hunk 6]
```diff
@@ -101,13 +105,43 @@ where
...
         let gas_tracker =
             GasTracker::new(Gas::new(gas_limit), Gas::zero(), machine.context().tracing);
 
+        let state_access_tracker = StateAccessTracker::new(machine.context().price_list);
+
+        /* Origin */
+
+        // - We don't charge for looking up the message's origin (assumed to be done ahead of time
+        //   and in parallel.
+        // - We've already charged for updating the message's origin in preflight.
```
[the remark this answers is not in the document]
Comment (Stebalien) [reply fvm/src/call_manager/default.rs:123]: > Have we? I don't think we actually charge any gas for that.

https://github.com/filecoin-project/ref-fvm/pull/1521/files#diff-14a46a0508c911abeb198a19207d9682a2061c34d4fc7d2787e0f9eecf8714dcR509

[the remark this answers is not in the document]
Comment (Stebalien) [reply fvm/src/call_manager/default.rs:123]: (i.e., I've rolled it into the message inclusion cost)

File: fvm/src/call_manager/default.rs [hunk 7]
```diff
@@ -313,8 +350,6 @@ where
         actor_id: ActorID,
         delegated_address: Option<Address>,
     ) -> Result<()> {
-        let start = GasTimer::start();
```
[the remark this answers is not in the document]
Comment (Stebalien) [reply fvm/src/call_manager/default.rs:316]: I'd kind of like to leave it for benchmarking. However, it only makes sense for benchmarks that don't require lotus (at least at the moment).

File: fvm/src/call_manager/default.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (arajasek) [reply fvm/src/call_manager/default.rs:123]: aaah, i see

File: fvm/src/call_manager/default.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (arajasek) [reply fvm/src/call_manager/default.rs:316]: Sure, SGWM.

File: fvm/src/call_manager/default.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (arajasek) [reply fvm/src/call_manager/default.rs:395]: ah, yes, agreed

File: fvm/src/call_manager/default.rs [hunk 8]
```diff
@@ -359,6 +395,89 @@ where
...
+
+        let mut to_actor = self.get_actor(to)?.ok_or_else(
+            || syscall_error!(NotFound; "transfer recipient {to} does not exist in state-tree"),
+        )?;
+
+        from_actor.deduct_funds(value)?;
+        to_actor.deposit_funds(value);
+
+        self.set_actor(from, from_actor)?;
+        self.set_actor(to, to_actor)?;
```
[the remark this answers is not in the document]
Comment (arajasek) [reply fvm/src/call_manager/default.rs:485]: I think this is perfectly fine as-is, nothing confusing or concerning here.

**Review by arajasek — APPROVED** [review]

**Pull request merged**